ui/tests/unit/routes/vault/cluster/oidc-callback-test.js

```diff
@@ -0,0 +1,184 @@
+    this.route.paramsFor = (path) => {
+      if (path === 'vault.cluster') return { namespaceQueryParam: 'admin' };
+      return {
+        auth_path: 'oidc-dev',
+        state: this.state('admin/child-ns'),
+        code: this.code,
+      };
+    };
+    this.route.afterModel();
+    assert.propContains(
+      this.windowStub.lastCall.args[0],
+      {
+        path: 'oidc-dev',
+        namespace: 'admin/child-ns',
+        state: this.state(),
+      },
+      'state ns takes precedence, state no longer has ns query'
+    );
+  });
+
+  test(`it uses namespace from namespaceQueryParam when state does not include: ',ns=some-namespace'`, function (assert) {
+    this.routeName = 'vault.cluster.oidc-callback';
+    this.route.paramsFor = (path) => {
+      if (path === 'vault.cluster') return { namespaceQueryParam: 'admin' };
+      return {
+        auth_path: this.path,
+        state: this.state(),
+        code: this.code,
+      };
+    };
+    this.route.afterModel();
+    assert.propContains(
+      this.windowStub.lastCall.args[0],
+      {
+        path: this.path,
+        namespace: 'admin',
+        state: this.state(),
+      },
+      'namespace is from cluster namespaceQueryParam'
+    );
+  });
+
+  test('it uses ns param from state when no namespaceQueryParam from cluster', function (assert) {
+    this.routeName = 'vault.cluster.oidc-callback';
+    this.route.paramsFor = (path) => {
+      if (path === 'vault.cluster') return { namespaceQueryParam: '' };
+      return {
+        auth_path: this.path,
+        state: this.state('ns1'),
+        code: this.code,
+      };
+    };
+    this.route.afterModel();
+    assert.propContains(
+      this.windowStub.lastCall.args[0],
+      {
+        path: this.path,
+        namespace: 'ns1',
+        state: this.state(),
+      },
+      'it strips ns from state and uses as namespace param'
+    );
+  });
+
+  test('the afterModel hook returns when both cluster and route params are empty strings', function (assert) {
+    this.routeName = 'vault.cluster.oidc-callback';
+    this.route.paramsFor = (path) => {
+      if (path === 'vault.cluster') return { namespaceQueryParam: '' };
+      return {
+        auth_path: '',
+        state: '',
+        code: '',
+      };
+    };
+    this.route.afterModel();
+    assert.propContains(
+      this.windowStub.lastCall.args[0],
+      {
+        path: '',
+        state: '',
+        code: '',
+      },
+      'model hook returns with empty params'
+    );
+  });
+
+  test('the afterModel hook returns when state param does not exist', function (assert) {
+    this.routeName = 'vault.cluster.oidc-callback';
+    this.route.paramsFor = (path) => {
+      if (path === 'vault.cluster') return { namespaceQueryParam: '' };
+      return {
+        auth_path: this.path,
+      };
+    };
+    this.route.afterModel();
+    assert.propContains(
+      this.windowStub.lastCall.args[0],
+      {
+        code: undefined,
```

Comment:
I think leaving these as undefined might lead to some weird behavior on the `postMessage` receiver side, could we default to empty string if these don't exist? A bit more verbose but something like: 
```
     let queryParams = { source, path: path || '', code: code || '', state: state || '' }; 
```
Maybe as a follow-on, because I think that's not likely a scenario we'll get into in real life